src/main/java/forestry/core/recipes/RecipeUtil.java

```diff
@@ -132,28 +131,24 @@ private static ItemStack getCraftingEquivalent(NonNullList<ItemStack> stockCopy,
 		return ForgeRegistries.RECIPES.getValues().stream().filter(recipe -> recipe.matches(inventory, world)).collect(Collectors.toList());
 	}
 
-	public static void addRecipe(String recipeName, Block block, Object... obj) {
-		addRecipe(recipeName, new ItemStack(block), obj);
+	public static void addRecipe(String moduleUID, Block block, Object... obj) {
+		addRecipe(moduleUID, new ItemStack(block), obj);
 	}
 
-	public static void addRecipe(String recipeName, Item item, Object... obj) {
-		addRecipe(recipeName, new ItemStack(item), obj);
+	public static void addRecipe(String moduleUID, Item item, Object... obj) {
+		addRecipe(moduleUID, new ItemStack(item), obj);
 	}
 
-	public static void addRecipe(String recipeName, ItemStack itemstack, Object... obj) {
-		ShapedRecipeCustom recipe = new ShapedRecipeCustom(itemstack, obj);
```

Comment:
Hmm, I don't think the current JSONs allow moving smaller recipes in the crafting grid (although they do use oredict). I'm hopeful there might be a better way to do this in the file rather than register one recipe for each position in the crafting grid though. It might require a custom loader.

Reply:
Allow moving smaller recipes in the crafting grid works with vanilla recipes as well. The sense behind `ShapedRecipeCustom` is another one. In 1.7.10 the recipe never checked the nbt of the ingredients so we had to add a custom recipe for the farm.

Reply:
I atleast think that that is the sense behind it. @mezz please correct me if I am wrong.

Reply:
Ah OK I see what you mean. I had a look and it seems you can get around this. It looks like CoFH also has recipes involving NBT tags (liquids in tanks and owners on devices) and it preserves them like this:
https://github.com/CoFH/CoFHCore/blob/dev/src/main/java/cofh/core/util/crafting/ShapedUpgradeRecipeFactory.java
https://github.com/CoFH/ThermalExpansion/blob/dev/src/main/resources/assets/thermalexpansion/recipes/_factories.json
https://github.com/CoFH/ThermalExpansion/blob/dev/src/main/resources/assets/thermalexpansion/recipes/satchel_1.json
I just need to implement that.

Reply:
There should be a IngredientNBT.

Reply:
Seems like it should be. Now I have to just understand how the system works properly. Would have helped if there was better documentation...

Reply:
Seems like NBT is already built in ?
https://github.com/temp1011/ForestryMC/blob/3083abad1ef24568ee840c802bf992a3ce1f7a0c/src/main/java/forestry/core/recipes/json/RecipeConverter.java#L189-L192